### PR TITLE
HPCC-29121 Track and report Roxie agent time spent waiting for a free thread

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/ccd/ccdqueue.cpp [hunk 1]
```diff
@@ -1573,6 +1594,7 @@ class CRoxieWorker : public CInterface, implements IPooledThread
                 {
                     for (unsigned subChannel = 0; subChannel < mySubChannel; subChannel++)
                         noteNodeSick(header.subChannels[subChannel]);
+                    logctx.setStatistic(StTimeIBYTIDelay, (msTick()-start)*1000000);
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp:1597]: start may not be initialised.  Could already cause problems with the tracing.

Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp:1597]: I keep thinking it is a shame we're not using nsTick().  Not a change to make at the same time.
Could use for StTimeAgentQueue, (which is also interesting on a sub-ms timing level).

**Review by ghalliday — COMMENTED** [review]
Generally looks ok, one problem and a shame about the accuracy.  I agree the ibyti on retry looks like it needs revisiting, but a separate PR.

File: roxie/ccd/ccdqueue.cpp (hunk 1 above)
Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp:1597]: Can do...

Comment (mckellyln) [inline roxie/ccd/ccdqueue.cpp:1597]: And perhaps hoist msTick()-start ?

**Review by mckellyln — COMMENTED** [review]
Looks fine, except for fixing start and yes good idea to use nsTick() instead of having to *1000000 often.

File: roxie/ccd/ccdqueue.cpp [hunk 2]
```diff
@@ -1725,6 +1749,10 @@ class CRoxieWorker : public CInterface, implements IPooledThread
                     if (next)
                     {
                         logctx.set(next);
+                        logctx.setStatistic(StTimeAgentQueue, (next->queryEnqueuedTimeStamp()-msTick())*1000000);
```
Comment (richardkchapman) [inline roxie/ccd/ccdqueue.cpp]: Should be mstick()-queryEnqueued()

**Comment by richardkchapman** [comment]
Changed to track nanoSeconds, and addressed other comments from review.

**Review by ghalliday — APPROVED** [review]
@richardkchapman looks useful. Please squash.

**Review by mckellyln — APPROVED** [review]
Looks good

**Pull request merged**